File: 2646-minimize-the-total-price-of-the-trips/2646-minimize-the-total-price-of-the-trips.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {number[]} price
 * @param {number[][]} trips
 * @return {number}
 */
var minimumTotalPrice = function(n, edges, price, trips) {
  const graph = Array.from({ length: n }, () => []);
  const count = new Array(n).fill(0);
  
  edges.forEach(([u, v]) => {
    graph[u].push(v);
    graph[v].push(u);
  })
  
  trips.forEach(([start, end]) => dfs1(start, -1, end, count, graph));
  
  return Math.min(...dfs2(0, -1, count, price, graph));
};

const dfs1 = (u, father, end, count, graph) => {
  if (u === end) {
    count[u]++;
    return true;
  }
  
  for (let v of graph[u]) {
    if (v !== father && dfs1(v, u, end, count, graph)) {
      count[u]++;
      return true;
    }
  }
  
  return false;
}

const dfs2 = (u, father, count, price, graph) => {
  let full = count[u] * price[u];
  let half = Math.floor(full / 2);
  
  for (let v of graph[u]) {
    if (v !== father) {
      const [ff, hh] = dfs2(v, u, count, price, graph);
      full += Math.min(ff, hh);
      half += ff;
    }
  }
  
  return [full, half];
}

File: 2646-minimize-the-total-price-of-the-trips/2646-minimize-the-total-price-of-the-trips.py
class Solution:
    def minimumTotalPrice(self, n: int, edges: List[List[int]], price: List[int], trips: List[List[int]]) -> int:
        graph = [[] for _ in range(n)]
        count = [0] * n
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
            
        for start, end in trips:
            def dfs(u, father): 
                if u == end:
                    count[u] += 1
                    return True
                for v in graph[u]:
                    if v != father and dfs(v, u):
                        count[u] += 1
                        return True
                return False
            
            dfs(start, -1)
            
        
        def dfs(u, father):
            full = count[u] * price[u]
            half = full // 2
            for v in graph[u]:
                if v != father:
                    ff, hh = dfs(v, u)
                    full += min(ff, hh)
                    half += ff
            return full, half
          
        return min(dfs(0, -1))